devicetypes/shinasys/sihas-zigbee-metering-plug.src/sihas-zigbee-metering-plug.groovy

```diff
@@ -0,0 +1,175 @@
+ *	in compliance with the License. You may obtain a copy of the License at:
+ *
+ *		http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *	Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *	on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *	for the specific language governing permissions and limitations under the License.
+ *
+ */
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition (name: "SiHAS Zigbee Metering Plug", namespace: "shinasys", author: "SHINA SYSTEM", mnmn: "SmartThingsCommunity", ocfDeviceType: "oic.d.smartplug", vid: "12d61425-2258-376a-beee-7a69fbc0d9fe") {
+		capability "Energy Meter"
+		capability "Power Meter"
+		capability "Refresh"
+		capability "Health Check"
+		capability "Sensor"
+		capability "Configuration"
+		capability "Voltage Measurement"
+		capability "afterguide46998.currentMeasurement"
+		capability "afterguide46998.frequencyMeasurement"
+		capability "afterguide46998.powerfactorMeasurement"
+		capability "Temperature Measurement"
+		capability "Switch"
+
+		fingerprint profileId: "0104", manufacturer: "ShinaSystem", model: "CCM-300Z2", deviceJoinName: "SiHAS Outlet" // SIHAS Zigbee Metering Plug 01 0104 0000 01 06 0000 0004 0003 0006 0B04 0702 02 0004 0019		
+	}
+}
+
+def getATTRIBUTE_READING_INFO_SET() { 0x0000 }
+def getATTRIBUTE_HISTORICAL_CONSUMPTION() { 0x0400 }
+def getATTRIBUTE_ACTIVE_POWER() { 0x050B }
+def getATTRIBUTE_FREQUENCY() { 0x0300 }
+def getATTRIBUTE_VOLTAGE() { 0x0505 }
+def getATTRIBUTE_CURRENT() { 0x0508 }
+def getATTRIBUTE_POWERFACTOR() { 0x0510 }
+def getTEMPERATURE_MEASUREMENT_MEASURED_VALUE_ATTRIBUTE() { 0x0000 }
+
+def convertHexToInt24Bit(value) {
+	int result = zigbee.convertHexToInt(value)
+	if (result & 0x800000) {
+		result |= 0xFF000000
+	}
+	return result
+}
+
+def parse(String description) {
+	log.debug "description is $description"
+	def event = zigbee.getEvent(description)
+	def descMap = zigbee.parseDescriptionAsMap(description)
+
+	if (event) {
+		log.info "event enter:$event"
+		if (event.name == "switch") {
+			return sendEvent(event)
+		} else if (event.name == "temperature") {
+			return sendEvent(event)
+		}
+	}
+
+	if (descMap) {
+		List result = []
+		log.debug "Desc Map: $descMap"
+
+		List attrData = [[clusterInt: descMap.clusterInt ,attrInt: descMap.attrInt, value: descMap.value, isValidForDataType: descMap.isValidForDataType]]
+		descMap.additionalAttrs.each {
+			attrData << [clusterInt: descMap.clusterInt, attrInt: it.attrInt, value: it.value, isValidForDataType: it.isValidForDataType]
+		}
+		attrData.each {
+			def map = [:]
+			if (it.isValidForDataType && (it.value != null)) {
+				if (it.clusterInt == zigbee.SIMPLE_METERING_CLUSTER && it.attrInt == ATTRIBUTE_HISTORICAL_CONSUMPTION) {
+					log.debug "meter"
+					map.name = "power"
+					map.value = convertHexToInt24Bit(it.value)/powerDivisor
```

Comment:
Why this value is converted to Int24Bit? Groovy would not handle it if value were converted to Int?

Reply:
This is also the same reason as PowerFactor. The type is int24.